x/epochs/abci.go

```diff
@@ -0,0 +1,81 @@
+package epochs
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/c-osmosis/osmosis/x/epochs/keeper"
+	"github.com/c-osmosis/osmosis/x/epochs/types"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// BeginBlocker of epochs module
+func BeginBlocker(ctx sdk.Context, k keeper.Keeper) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
+
+	// Note: epoch length must be more than 2 blocks, as the implementation requires epoch_endblock and epoch_startblock are separate
+	// epoch_startblock(n+1) = epoch_endblock(n) + 1
+	k.IterateEpochInfo(ctx, func(index int64, epochInfo types.EpochInfo) (stop bool) {
+		if !epochInfo.EpochCountingStarted { // epoch counting not started
+			if epochInfo.StartTime.Before(ctx.BlockTime()) { // check time is good to start
+				epochInfo.EpochCountingStarted = true
+				epochInfo.CurrentEpochEnded = false
+				epochInfo.CurrentEpoch = 0
+				epochInfo.CurrentEpochStartTime = ctx.BlockTime()
+				ctx.EventManager().EmitEvent(
+					sdk.NewEvent(
+						types.EventTypeEpochStart,
+						sdk.NewAttribute(types.AttributeEpochNumber, fmt.Sprintf("%d", epochInfo.CurrentEpoch)),
+						sdk.NewAttribute(types.AttributeEpochStartTime, fmt.Sprintf("%d", epochInfo.CurrentEpochStartTime.Unix())),
+					),
+				)
+			}
+		} else if epochInfo.CurrentEpochEnded { // epoch ended in last block
+			epochInfo.CurrentEpochEnded = false
+			epochInfo.CurrentEpoch = epochInfo.CurrentEpoch + 1
+			epochInfo.CurrentEpochStartTime = ctx.BlockTime()
```

Comment:
So something thats being done here is we are actually taking drift in epoch start times. (What I called 'block time drift' here https://github.com/osmosis-labs/osmosis/pull/108/files#r624009821)

I'm not sure if we want to be doing that, versus instead making `CurrentEpochStartTime` equal to the exact time it would have began at. If w

/cc @sunnya97

Reply:
Actually on further thought, fixing the block time drift would require some changes to how we flag when to begin the next epoch. I think this is important to do, but I also think we can do it post-launch.

So I'm happy to leave the block time drift for now, since for our usecase its not going to be a big deal. (We'll shift around 20 seconds at most per week I think)

And then we relegate solving this to #postlaunch

Reply:
Added this on future improvement section of spec

Reply:
Issue opened